### PR TITLE
CST 124472 E2E Testing Rewrite of your-claims

### PR DESCRIPTION
**Pull request by williamphelps13**
# PR Summary: CST E2E Testing Rewrite - Your Claims Page

## Summary

- E2E testing rewrite to cover `/your-claims` page
- I am on BMT1 which is the code owner of the CST application

---

## E2E Tests - Before

The following test cases were covered by the existing E2E tests on main branch:

### Claims List (`00.claims-list.cypress.spec.js`)

- **Tests consolidated claim functionality**
  - Page title: "Check your claim, decision review, or appeal status | Veterans Affairs"
  - Breadcrumbs visible with 2 items
  - Current page breadcrumb has `aria-current="page"`
  - Page heading: "Check your claim, decision review, or appeal status"
  - First claim card header: "Claim for compensation"
  - Card status: "Moved to this step on October 31, 2016"
  - Clicking claim link navigates to `/your-claims/189685/status`
  - Expands all accordions

- **Displays server-generated claim titles in list view**
  - Claim card displays server's `displayTitle` field

- **Falls back to client-side titles when server fields absent**
  - Claim card shows client-generated title when `displayTitle` is undefined

- **Display document upload status alerts (when feature toggle enabled)**
  - No alerts shown when feature toggle is disabled
  - When toggle enabled and failed evidence submissions exist:
    - Shows alert for "Claim for disability compensation": "We need you to resubmit files"
    - Shows alert for "Edith Nourse Rogers STEM Scholarship application": "We need you to resubmit files"
    - Shows alert for "Supplemental claim": "We need you to resubmit files"

### Empty Claims List (`00.claims-list-empty.cypress.spec.js`)

- **Verifies functionality with an empty list**
  - Page loads successfully
  - Empty state alert: "You do not have any submitted claims"
  - Expands all accordions

### Breadcrumb (`00.claims-list-breadcrumb.cypress.spec.js`)

- **Verifies breadcrumb functionality** *(SKIPPED - test does not run)*
  - Second breadcrumb has `aria-current="page"` and text "Check your claims and appeals"
  - Calls undefined mobile breadcrumb method (test broken)

---

## E2E Tests - After

The following test cases are covered by the new E2E tests:

### Page Structure (`your-claims.cypress.spec.js`)

- Page title: "Check your claim, decision review, or appeal status | Veterans Affairs"
- Display breadcrumbs
  - "VA.gov home" link with `href="/"`
  - "Check your claims and appeals" link with `href="#content"`
- Display page heading
- **[NEW]** Display "On this page" navigation component
  - "Your claims, decision reviews, or appeals" link
  - "Your claim letters" link
  - "What if I can't find my claim, decision review, or appeal?" link
  - "Your travel claims" link
  - "Additional services" link
- Display card list section heading
- **[NEW]** Display "why we combine claims" additional info accordion (collapsed state)
- **[NEW]** Display additional info content when clicked
- Display no claims message when empty
- **[NEW]** Display claim letters section
  - Section heading
  - "Download your VA claim letters" link with correct href
  - Descriptive text
- **[NEW]** Display "What if I can't find" section
  - Section heading
  - Descriptive text about processing time
- **[NEW]** Display travel claims section
  - Section heading
  - "Review and file travel claims" link with correct href
  - Descriptive text
- **[NEW]** Display additional services section
  - "Change your address and other contact information" link
  - "Change your legal name on file with VA" link
  - "Get help filing your VA claim, decision review, or appeal" link
- **[NEW]** Display need help section
  - "Need help?" heading in shadow DOM
  - VA benefits hotline text

### Claim Cards (`claim-cards.cypress.spec.js`)

- **[ENHANCED]** Display completed compensation claim
  - Heading with claim type and received date
  - Phase status: "Step 5 of 5: Closed"
  - "Moved to this step on..." date
  - Details link with correct href
- **[ENHANCED]** Display in-progress compensation claim
  - Heading with "In Progress" prefix
  - Phase status: "Step 1 of 5: Claim received"
  - Details link

**Claim Type Coverage:**
- Display claim for disability compensation
- **[NEW]** Display claim for expenses related to death or burial
- **[NEW]** Display request to add or remove a dependent
- **[NEW]** Display claim for Survivors Pension
- **[NEW]** Display claim for Dependency and Indemnity Compensation
- **[NEW]** Display claim for Veterans Pension
- **[NEW]** Display claim for pension

**Communication Notifications:**
- **[ENHANCED]** Display development letter notification
  - "We sent you a development letter" text
- **[ENHANCED]** Display decision letter notification
  - "You have a decision letter ready" text

**5-Phase System (Default):**
- **[NEW]** Display Step 1 of 5: Claim received
- Display Step 2 of 5: Initial review
- **[NEW]** Display Step 3 of 5: Evidence gathering, review, and decision
- **[NEW]** Display Step 4 of 5: Preparation for notification
- **[NEW]** Display Step 5 of 5: Closed

**8-Phase System (Feature Flag Enabled):**
- **[NEW]** Display Step 1 of 8: Claim received
- **[NEW]** Display Step 2 of 8: Initial review
- **[NEW]** Display Step 3 of 8: Evidence gathering
- **[NEW]** Display Step 4 of 8: Evidence review
- **[NEW]** Display Step 5 of 8: Rating
- **[NEW]** Display Step 6 of 8: Preparing decision letter
- **[NEW]** Display Step 7 of 8: Final review
- **[NEW]** Display Step 8 of 8: Claim decided

**Document Alerts:**
- **[ENHANCED]** Display documents needed alert
  - "We requested more information from you:" text

### Appeal Cards (`appeal-cards.cypress.spec.js`)

**Appeal Types:**
- **[NEW]** Display legacy appeal card
  - Heading: "Disability compensation appeal Received on..."
  - "Issue on appeal" / "Issues on appeal" label
  - Issue description (e.g., "Tinnitus")
  - Status text
  - "Last updated:" date
  - Details link with correct href to `/track-claims/appeals/[id]/status`
- **[NEW]** Display supplemental claim card
  - Heading: "Supplemental claim for [program area] Received on..."
  - "Issue on review" / "Issues on review" label
  - Issue description
  - Status text
  - "Last updated:" date
  - Details link
- **[NEW]** Display higher-level review card
  - Heading: "Higher-level review for [program area] Received on..."
  - "Issue on review" / "Issues on review" label
  - Issue description
  - Status text
  - "Last updated:" date
  - Details link
- **[NEW]** Display appeal (AMA) card
  - Heading: "Disability compensation appeal Received on..."
  - "Issue on appeal" / "Issues on appeal" label
  - Issue description
  - Status text
  - "Last updated:" date
  - Details link

**Program Area Coverage (9 areas × 2 card types = 18 tests):**
- **[NEW]** Display compensation appeal
- **[NEW]** Display supplemental claim for compensation
- **[NEW]** Display pension appeal
- **[NEW]** Display supplemental claim for pension
- **[NEW]** Display insurance appeal
- **[NEW]** Display supplemental claim for insurance
- **[NEW]** Display loan guaranty appeal
- **[NEW]** Display supplemental claim for loan guaranty
- **[NEW]** Display education appeal
- **[NEW]** Display supplemental claim for education
- **[NEW]** Display vocational rehabilitation and employment appeal
- **[NEW]** Display supplemental claim for vocational rehabilitation and employment
- **[NEW]** Display health care (medical) appeal
- **[NEW]** Display supplemental claim for health care
- **[NEW]** Display burial benefits appeal
- **[NEW]** Display supplemental claim for burial benefits
- **[NEW]** Display fiduciary appeal
- **[NEW]** Display supplemental claim for fiduciary

**Issues Display:**
- **[NEW]** Display "Issue on appeal" for single issue (appeal type)
- **[NEW]** Display "Issues on appeal" for multiple issues (appeal type)
- **[NEW]** Display "Issue on review" for single issue (review type)
- **[NEW]** Display "Issues on review" for multiple issues (review type)

**Edge Cases:**
- **[NEW]** Do not display issues section when description is missing

### STEM Claim Cards (`stem-cards.cypress.spec.js`)

- **[NEW]** Display denied STEM claim
  - Title: "Edith Nourse Rogers STEM Scholarship application"
  - "Received on" date
  - Status: "Denied"
  - "Last updated on:" date
  - Details link with correct href to `/track-claims/your-stem-claims/[id]/status`
- **[NEW]** Do not display non-denied STEM claim
  - When `automatedDenial` is false, card should not appear

**Upload Error Alerts:**
- **[NEW]** Display upload error alert for failed submissions within last 30 days
  - Alert text: "We need you to resubmit files for this claim."
- **[NEW]** Do not display upload error alert for failed submissions older than 30 days
  - No error alert should appear

### Error States (`your-claims-unavailable.cypress.spec.js`)

- **[NEW]** Display claims and appeals unavailable alert
  - When both `/v0/benefits_claims` and `/v0/appeals` return 500 error
  - Heading: "Claim and Appeal status is unavailable"
  - Instructional text: "VA.gov is having trouble loading claims and appeals information at this time. Check back again in an hour."
- **[NEW]** Display claims unavailable alert
  - When `/v0/benefits_claims` returns 500 error (appeals load successfully)
  - Heading: "Claim status is unavailable"
  - Text includes note that appeals are still available
- **[NEW]** Display appeals unavailable alert
  - When `/v0/appeals` returns 500 error (claims load successfully)
  - Heading: "Appeal status is unavailable"
  - Text includes note that claims are still available

### Loading States (`loading.cypress.spec.js`)

- **[NEW]** Display required login view loader
  - Loader visible during `/v0/user` authentication (delayed 500ms)
  - Loader (`[data-testid="required-login-view-loader"]`) visible then disappears
  - Page heading appears after loader
- **[NEW]** Display downtime notification loader
  - Loader visible during `/v0/maintenance_windows` check (delayed 1500ms)
  - Loader (`[data-testid="downtime-notification-loader"]`) visible then disappears
  - Page heading appears after loader
- **[NEW]** Display feature flags loader
  - Loader visible during `/v0/feature_toggles` fetch (delayed 1500ms)
  - Loader (`[data-testid="feature-flags-loader"]`) visible then disappears
  - Page heading appears after loader

---

### Test Coverage Statistics

| Metric | Before | After | Change |
|--------|--------|-------|--------|
| Total tests | 6 active (7 total, 1 skipped) | 75 | +1,150% |
| Claim types tested | 2 | 7 | +250% |
| Appeal types tested | 0 | 4 | New |
| Program areas tested | 0 | 9 | New |
| Phase systems tested | Partial (5-phase) | Complete (5 & 8-phase) | Both systems |
| Error states tested | 0 | 3 | New |
| Loading states tested | 0 | 3 | New |

### New Coverage Areas (47+ tests marked [NEW])

**Page sections and navigation:**
- Page heading
- "On this page" navigation component (5 links)
- Combined claims accordion interaction (collapsed/expanded states)
- Claim letters section (heading, link, text)
- "What if I can't find" section
- Travel claims section
- Additional services section (3 links)
- Need help section

**Claim cards:**
- 5 additional claim types (pension, survivors, DIC, dependents, burial)
- 8-phase system (all 8 phases)
- 4 additional 5-phase tests (steps 1, 3, 4, 5)

**Appeal cards (27 new tests):**
- Appeal types
- 4 issues display variations

**STEM claim cards (4 new tests):**
- Denied STEM claim display
- Non-denied STEM claim (should not display)
- Upload error alerts with date logic (within 30 days / older than 30 days)

**Error states (3 new tests):**
- Both services unavailable
- Claims only unavailable
- Appeals only unavailable

**Loading states (3 new tests):**
- Required login view loader
- Downtime notification loader
- Feature flags loader

### Enhanced Coverage (5 tests marked [ENHANCED])

- Explicit distinction between completed vs in-progress claims
- Development letter notification (existed in fixtures but not explicitly tested)
- Decision letter notification (existed in fixtures but not explicitly tested)
- Documents needed alert (existed as upload status alert, now tested separately)
- Complete 5-phase system coverage (only partial before)

### Test Quality Improvements

**Modern test patterns:**
- ✅ DRY setup functions: `setupClaimCardsTest()`, `setupAppealCardsTest()`, etc.
- ✅ Test data factories: `createClaim()`, `createAppeal()`, `createStemClaim()`, `createEvidenceSubmission()`
- ✅ Accessibility testing: Every test includes `cy.axeCheck()`
- ✅ Shadow DOM testing: Proper `.shadow()` usage for web components
- ✅ User-focused assertions: Testing visible text with `.findByRole()`, `.findByText()` instead of implementation details
- ✅ Context grouping: Logical test organization with `context()` blocks for maintainability

**Test architecture:**
- Modular organization: Tests separated by component/functionality
- Reusable helpers: Factory functions for creating consistent test data
- Clear naming: Descriptive test names that explain what's being tested
- Independent tests: Each test can run in isolation with proper setup
- Commented documentation: Test data factories include comments on which fields affect display

### Best Practices Impact

**New E2E Testing Standards ([README](./README.md))**

The rewrite introduces comprehensive best practices aligned with Cypress and VA Platform standards:

**1. Test Isolation & State Management**
- **Before**: Tests used shared page objects and built up state through UI interactions
- **After**: Each test uses isolated setup functions and `cy.intercept()` to set state directly
- **Impact**: Tests run faster (no UI navigation overhead), and when one test fails, it doesn't cause unrelated tests to fail. Each test starts fresh, making it easier to pinpoint which feature actually broke.

**2. Accessibility-First Testing**
- **Before**: Tests had `cy.axeCheck()` but used CSS selectors (`.claim-list-item-header`)
- **After**: Tests use Testing Library queries (`findByRole`, `findByText`) that match how users interact with the page
- **Impact**: Tests now validate that screen reader users can find and interact with elements the same way sighted users can. If a test passes, we know both visual and assistive technology users can complete the task.

**3. Test Data Factories**
- **Before**: Tests used complex fixture files that were hard to understand and modify
- **After**: Factory functions (`createClaim()`, `createAppeal()`) with clear, documented parameters
- **Impact**: Developers can quickly test edge cases (like claims with missing data or specific dates) without maintaining large JSON files. Comments in the code explain which data fields control what users see, making it easier to add tests for new scenarios.

**4. Maintainability**
- **Before**: Monolithic test files with duplicated setup code
- **After**: Modular structure with DRY principles and reusable helpers
- **Impact**: New team members can find relevant tests faster. Adding tests for new features takes less time because setup code is reusable. When a test fails, the file name and test organization make it immediately clear which page feature is broken.

**Example Comparison:**

```javascript
// Before: Page object pattern with shared state
const trackClaimsPage = new TrackClaimsPage();  // Shared object used across tests
trackClaimsPage.loadPage(claimsList);           // Load entire fixture file
trackClaimsPage.checkClaimsContent();           // Hidden implementation details

// After: Isolated test with factory and accessibility-first
setupClaimCardsTest([createClaim({ status: 'COMPLETE' })]);  // Create exactly the data needed
cy.findByRole('heading', {                                   // Find by accessibility role
  name: 'Claim for compensation Received on January 1, 2025' // Match what users see/hear
});
cy.axeCheck();  // Verify accessibility compliance
```

**Documentation Impact:**
- E2E README now provides clear guidance for writing maintainable, accessible tests
- Team alignment on best practices for future test development
- Scalable patterns that work across the entire CST application

---

### Old Files (To be removed after standards solidify)

- ❌ `src/applications/claims-status/tests/e2e/00.claims-list-breadcrumb.cypress.spec.js` (1 test, skipped)
- ❌ `src/applications/claims-status/tests/e2e/00.claims-list.cypress.spec.js` (5 tests)
- ❌ `src/applications/claims-status/tests/e2e/00.claims-list-empty.cypress.spec.js` (1 test)

### New Test Files

```
src/applications/claims-status/tests/e2e/tests/
├── shared/
│   └── loading.cypress.spec.js                    (3 tests)
└── your-claims/
    ├── your-claims.cypress.spec.js                (13 tests)
    ├── claim-cards.cypress.spec.js                (25 tests)
    ├── appeal-cards.cypress.spec.js               (27 tests)
    ├── stem-cards.cypress.spec.js                 (4 tests)
    └── your-claims-unavailable.cypress.spec.js    (3 tests)
```

---

## Next Steps

**Note:** This retains the original tests for now as the standards and coverage for this rewrite are still being refined. The next part of this rewrite is to rewrite details tests which will result in most of the application being covered and therefore will solidify the new testing standards.

---

## Related issue(s)

- [#124472 Increase end-to-end test coverage for the /your-claims flow to strengthen reliability of the Claim Status Tool](https://github.com/department-of-veterans-affairs/va.gov-team/issues/124472)

### CODE REVIEW

File: src/applications/claims-status/tests/e2e/tests/shared/loading.cypress.spec.js [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+    cy.login();
+    cy.visit('/track-claims');
+
+    cy.findByTestId('feature-flags-loader').should('be.visible');
+
+    cy.findByTestId('feature-flags-loader').should('not.exist');
+    cy.findByRole('heading', {
+      name: 'Check your claim, decision review, or appeal status',
+    });
```
Comment (nathangthomas) [inline src/applications/claims-status/tests/e2e/tests/shared/loading.cypress.spec.js]: }).should('be.visible');

**Review by nathangthomas — CHANGES_REQUESTED** [review]
Awesome work here! I only have one requested change. Happy to talk about it in more detail if needed.

File: src/applications/claims-status/tests/e2e/tests/shared/loading.cypress.spec.js [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+              'appeals-status',
+            ],
+            va_profile: {
+              status: 'OK',
+            },
+          },
+        },
+        meta: { errors: null },
+      },
+    });
```
[the remark this answers is not in the document]
Comment (williamphelps13) [reply src/applications/claims-status/tests/e2e/tests/shared/loading.cypress.spec.js]: Fixed by: [cd44b71](https://github.com/department-of-veterans-affairs/vets-website/pull/40123/commits/cd44b71dd868240a8d5e56f2ce823d9506ac6bbc)

File: src/applications/claims-status/tests/e2e/README.md [hunk 3]
```diff
@@ -14,7 +14,7 @@ Source: [Cypress Best Practices](https://docs.cypress.io/app/core-concepts/best-
 2. Use `data-*` attributes to provide context to your selectors and isolate them from CSS or JS changes ([source](https://docs.cypress.io/app/core-concepts/best-practices#Organizing-Tests-Logging-In-Controlling-State))
    1. Never: `cy.get('.btn.btn-large').click()` - coupled to styling
    2. Depends: `cy.contains('Submit').click()` - coupled to text content
-   3. Always: `cy.get('[data-testid="submit"]').click()` - not coupled styling or content of an element
+   3. Always: `cy.get('[data-testid="submit"]').click()` - not coupled to styling or content of an element
```
Comment (nathangthomas) [inline src/applications/claims-status/tests/e2e/README.md]: Nice catch!

**Review by nathangthomas — APPROVED** [review]
Nice typo catches in the README. I read right over those and didn't even notice. Looks good to me!
